.github/workflows/node.js.yml

```diff
@@ -11,33 +10,18 @@ on:
 jobs:
   build:
 
-    runs-on: ubuntu-20.04
-
     strategy:
       matrix:
-        node-version: [10.x, 12.x, 14.x, 15.x]
+        node-version: [10.x, 12.x, 14.x, 15.x, 16.x]
+        os: [ubuntu-latest, windows-latest, macos-latest]
+
+    runs-on: ${{ matrix.os }}
+
     steps:
       - uses: actions/checkout@v2
       - name: Use Node.js ${{ matrix.node-version }}
         uses: actions/setup-node@v1
         with:
           node-version: ${{ matrix.node-version }}
-      - name: Cache pnpm modules
-        uses: actions/cache@v2
-        env:
-          cache-name: cache-pnpm-modules
-        with:
-          # pnpm cache files are stored in `~/.pnpm-store` on Linux/macOS
-          path: ~/.pnpm-store
-          key: ${{ runner.os }}-build-${{ env.cache-name }}-${{ matrix.node-version }}-${{ hashFiles('**/package.json') }}
-          restore-keys: |
-            ${{ runner.os }}-build-${{ env.cache-name }}-${{ matrix.node-version }}-
-      - name: Install pnpm
-        uses: pnpm/action-setup@v1.2.1
-        with:
-          version: 5.18.9
-          run_install: true
-      - name: Lint ${{ matrix.node-version }}
-        run: npm run lint
       - name: Test ${{ matrix.node-version }}
-        run: npm run test:bare
+        run: npm install && npm run build
```

Comment:
This should probably be `npm ci` since we've gone to the trouble of not caching anything, installing all of the dev tools, and the pain of package-lock.  This can be fixed later, though.

Reply:
the reason it is traditional to not use `npm ci` in a situation like this is that `npm ci` is a newcomer to the node world - it was only introduced in node 5.7, which is more than halfway into node's lifecycle, and long after it had continuous integration norms.

`npm ci` is meant to be used when the ci build is different than the local build.  when a node programmer sees that `npm install && npm run build` - the default for their own local setup - is being used to test in ci, they learn quite a bit about the expectations of strength of their local test set.

(i don't actually understand how package-lock is relevant here, so maybe i'm missing something)

Reply:
`npm ci` requires a package-lock, and I'm still salty about having to maintain what I think of as a useless file (for a package with no dependencies), but I'm willing to do it since there's some potential theoretical benefit to it.  I'm looking for any real benefit to maintaining it so I stop feeling salty. :)

Note: that's all about my feelings about myself, and unrelated to the technical solution we pick.  It's just background.

We don't have to move to `npm ci`

Reply:
> I'm still salty about having to maintain what I think of as a useless file (for a package with no dependencies), but I'm willing to do it since there's some potential theoretical benefit to it. I'm looking for any real benefit to maintaining it so I stop feeling salty. :)

There is a benefit.  It's ridiculously niche, but it has burned me once in the real world.  There is a second theoretical benefit that I've never seen personally.

&nbsp;

# The thing I've seen

Node sometimes builds non-node services.  Those things are frequently built through tools like `node-gyp`.  Those occasionally interact with operating system level package managers.

Back in yon paleozoic era, when we still used things like `nightmare` to manage browser testing, it was reasonably common to have both `nightmare` and `selenium` in place, because you could do most of your acceptance tests in `nightmare` with 1/10 the effort of `selenium`, but since that was electron and locked you into current-`chrome`, you could also do your horror path stuff in `selenium` and have full browser backversion testing.

`selenium` requested the `cairo` (an svg renderer) version in the package manager; `electron` set a floor, instead.  `electron` had a several years newer version than `selenium` did.

We were image testing SVG.  The version of `cairo` in use determined whether several things we needed, particularly miter joins on path endcaps, actually worked correctly.

As a result, if you installed `electron` first, you'd get the version of `cairo` that had been modern when your chosen `electron` distro was used, which `selenium` would happily use

If on the other hand you installed `selenium` first, you'd get the old version, and then the supporting tools around `freetype` would cause a version lock (think `anaconda` requirements,) meaning that when it upgraded, it'd actually get the correct modern version, instead of whatever electron *thought* was the modern version

We had a significantly obsoleted electron version because one of our customers was struggling to update their SSL ciphers because they in turn had some customers running on like windows negative four or something

Therefore, depending on what order you installed the two non-node things, you'd get significantly different tooling behavior from your underlying support libraries

It took us three days to figure out what the hell was wrong.  Once we did, we just installed the libraries before doing node anything, and it went away

However if you're distributing something where a user needs to rely on `node-gyp` or its demon bretheren, order of construction can be important

&nbsp;

# The thing I haven't seen

Rumor has it an equivalent problem can exist inside of javascript node dependencies.

I've never understood how, or heard a coherent explanation, but the author of babel maintains this to be true, so I believe it.

I have never seen it and cannot explain.

Reply:
wrt `npm ci`, i'm probably actually going to move to it later, to give you the variant behavior you want for eslint

that's an appropriate time because at that point ci really is behaving differently than local

i was just explaining why i hadn't yet done that